### PR TITLE
add dosctings, pathlib, typing, uppercase constants etc. to few base files: WIP

### PR DESCRIPTION
**Pull request by marian-code**
I am pushing this PR as a follow up on dicussion #321 .

Some of the base modules where modified and also setup.py script

The list of changes:

* some basic code cleanup, formatting and refactoring where suitable with the aim to be more pythonic
* added detailed type hints
* ported parts of the code that handle paths from strings to `pathlib.Path` as this is the prefered way to interact with the filesystem in python 3.4+  and is a lot less error prone
* constants are now typed in uppercase to be easily recognizable as is the recommended naming convention in python
* added docstrings to all public methods and classes and to some private ones too
* removed old python 2 type formatting using `%` in favor of f-strings - this is the recommended way in python 3.6+ and again is a lot more readable and less error prone
* removed a few unused imports
* cleanup in setup.py and port to pathlib
* added some TODOs with ideas for improvement
* removed m2r from `setup.py` because it makes no sense to convert to `rst` format when `long_description_content_type=text/markdown`

I hope this is OK and if it is to your liking I can push other changes that help with the rewrite.

### CODE REVIEW

**Comment by marian-code** [comment]
There seems to be weird error in tests.

pip install with skbuild runs OK but tests throw error:

```bash
tensorflow.python.framework.errors_impl.NotFoundError: /opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/deepmd/libop_abi.so: undefined symbol: _ZNK10tensorflow8OpKernel11TraceStringERKNS_15OpKernelContextEb
```
as if though `libop` was not compiled right. I am not very proficient with `C/C++` so I am a bit uncertain how to debug this.
I am able to run these test locally without problems on python=3.8, gcc=6.3 and tf=2.4.1 without any errors.

File: .github/workflows/test_python.yml [hunk 1]
```diff
@@ -59,5 +59,6 @@ jobs:
     - run: |
         sudo apt update
         sudo apt install gcc-${{ matrix.gcc }} g++-${{ matrix.gcc }}
+    - run: pip install tensorflow
```
Comment (amcadmus) [inline .github/workflows/test_python.yml:62]: I feel this line triggered the failure in unittests.
@njzjz do you have any hint?

Comment (njzjz) [reply .github/workflows/test_python.yml:62]: let me check...

Comment (njzjz) [reply .github/workflows/test_python.yml:62]: Yes. And for original script, I forgot to add environment variables...

File: setup.py [hunk 2]
```diff
@@ -1,95 +1,108 @@
...
+    if cmake_version in SpecifierSet("<3.7"):
+        setup_requires.append("cmake")
 
+Path("deepmd").mkdir(exist_ok=True)
 
 setup(
     name="deepmd-kit",
-    setup_requires=setup_requires,
-    use_scm_version={'write_to': 'deepmd/_version.py'},
+    setup_requires=INSTALL_REQUIRES,
```
Comment (njzjz) [inline setup.py:71]: maybe broken here...
```suggestion
    setup_requires=setup_requires,
```

**Comment by njzjz** [comment]
Ok, it turns green.

File: .github/workflows/test_python.yml (hunk 1 above)
Comment (marian-code) [reply .github/workflows/test_python.yml]: seems like I was incorrectly trying to solve the lack of environment variables by installation of TF from pip

**Comment by marian-code** [comment]
Is it OK if I push other changes? Any suggestion what I should work on?

**Review by marian-code — COMMENTED** [review]
That was a stupid mistake, sorry you had to hunt that down.

**Comment by njzjz** [comment]
> Is it OK if I push other changes? Any suggestion what I should work on?

Yes

**Review by amcadmus — CHANGES_REQUESTED** [review]

File: deepmd/common.py [hunk 3]
```diff
@@ -1,218 +1,453 @@
...
+    ----------
+    root_dir : Union[str, Path]
+        starting directory
+
+    Returns
+    -------
+    List[str]
+        list of string pointing to system directories
+    """
+    matches = [d for d in Path(root_dir).rglob("*") if (d / "type.raw").is_file()]
```
[the remark this answers is not in the document]
Comment (marian-code) [reply deepmd/common.py:430]: It is fixed as requested, although I think using absolute paths is a bit safer.

**Review by amcadmus — APPROVED** [review]

**Pull request merged**